mqueue/rtqueue/rtqueue.go

```diff
@@ -0,0 +1,72 @@
+package rtqueue
+
+import (
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/mqueue"
+
+	"github.com/onflow/flow-go/utils/fifoqueue"
+)
+
+// Queue is a real-time queue which guarantees that items added to the queue
+// will immediately be available to be received over Recv channel.
+type Queue interface {
+	Add(item mqueue.Message)
+
+	Recv() <-chan mqueue.Message
+}
+
+type queue struct {
+	items      *fifoqueue.FifoQueue
+	in, out    chan mqueue.Message
+	shovelling *atomic.Bool
+}
+
+func New() Queue {
+	fifo, _ := fifoqueue.NewFifoQueue()
+	q := &queue{
+		items:      fifo,
+		in:         make(chan mqueue.Message),
+		out:        make(chan mqueue.Message),
+		shovelling: atomic.NewBool(false),
+	}
+	return q
+}
+
+// shovel is a goroutine that lives for as long as the queue is non-empty.
+func (q *queue) shovel() {
+
+	// shovel is always started when at least one item is being added
+	head := <-q.in
+
+	for {
+		select {
+		case tail := <-q.in:
+			// when a new item is added, re-enter this select with the same head
+			q.items.Push(tail)
+			continue
+		case q.out <- head:
+			// when the head of the queue is accepted, retrieve the next
+			// head before re-entering this select
+		}
+
```

Comment:
I think an element could get stuck:
* lets assume you have just pushed the last element into `q.out`: https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L48
* now, `q.items` is empty, so `ok` is `false`
* but before you get to https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L56, some thread pushes another element into `q.in`. 
    * The current go routine won't notice it and subsequently exit
    * but at the time when https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L64-L66 is evaluated, `q.shovelling` is still `false` so no new routine will be spawned to process the element ... 
   
   The culprit here is the non-atomicity of https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L53-L58 and  https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L64-L67 
     - you can circumvent this problem by not letting the go routine exit; then you arrive at solution https://github.com/onflow/flow-go/pull/460
     - alternatively, you can make these operations atomic by using a mutex. But then you don't need the `q.in` channel anymore and the thread executing https://github.com/onflow/flow-go/blob/2f8380cdae6c1f709179b24766eac884e5b6eb7c/mqueue/rtqueue/rtqueue.go#L63 could directly write to into the FifoQueue. The outbound channel is then a matter of taste: do I rather use a case statement to block? on channels or to I use the poker and poll the channels?